src/sdk/pynni/tests/test_tuner.py

```diff
@@ -17,107 +17,109 @@
+                    self.assertIsInstance(v, int)
+                    self.assertIn(v, item["_value"])
+                if item["_type"] == "uniform":
+                    self.assertIsInstance(v, float)
+                if item["_type"] in ("randint", "uniform", "quniform", "loguniform", "qloguniform"):
+                    self.assertGreaterEqual(v, item["_value"][0])
+                    self.assertLessEqual(v, item["_value"][1])
+                if item["_type"].startswith("q"):
+                    multiple = v / item["_value"][2]
+                    if item["_value"][0] + EPS < v < item["_value"][1] - EPS:
+                        self.assertAlmostEqual(int(round(multiple) + EPS), multiple)
+                if item["_type"] in ("qlognormal", "lognormal"):
+                    self.assertGreater(v, 0)
+
+    def search_space_test_all(self, tuner_factory, supported_types=None):
+        with open("assets/search_space.json", "r") as fp:
+            search_space_all = json.load(fp)
+        if supported_types is None:
+            supported_types = ["choice", "randint", "uniform", "quniform", "loguniform", "qloguniform",
+                               "normal", "qnormal", "lognormal", "qlognormal"]
+        full_supported_search_space = dict()
+        for single in search_space_all:
+            single_keyword = single.split("_")
+            single_search_space = {single: search_space_all[single]}
+            if any([t in single_keyword for t in supported_types]) and "fail" not in single_keyword:
+                try:
+                    # supports this key
+                    self.search_space_test_one(tuner_factory, single_search_space)
+                    full_supported_search_space.update(single_search_space)
+                except Exception as e:
+                    if "undefined" not in single_keyword:
+                        raise  # TODO: fix undefined
+                    logger.error(e)
+            else:
+                with self.assertRaises(Exception) as cm:
+                    self.search_space_test_one(tuner_factory, single_search_space)
+
+                logger.info("{}, {}, {}".format(tuner_factory, single, cm.exception))
+        logger.info("Full supported search space: {}".format(full_supported_search_space))
+        self.search_space_test_one(tuner_factory, full_supported_search_space)
+
+    def test_grid_search(self):
+        self.search_space_test_all(lambda: GridSearchTuner(),
+                                   supported_types=["choice", "randint", "quniform"])
+
+    def test_tpe(self):
+        self.search_space_test_all(lambda: HyperoptTuner("tpe"))
+
+    def test_random_search(self):
+        self.search_space_test_all(lambda: HyperoptTuner("random_search"))
+
+    def test_anneal(self):
+        self.search_space_test_all(lambda: HyperoptTuner("anneal"))
+
+    def test_smac(self):
+        # TODO: smac seems not clean and generates a lot of temporary files
```

Comment:
Can you fix these TODO at this PR or next PR?

Reply:
In this PR of course. I create this PR mainly to see pipeline results.